src/observers.ts

```diff
@@ -0,0 +1,251 @@
+  }
+  // there may be way to consolidate this
+  const isWithinEndRange = gateway.end === 0; // TODO: do we need this now that we got rid of hidden
+  return didStartBeforeEpoch && isWithinEndRange;
+}
+
+export function getEligibleGatewaysForEpoch({
+  epochStartHeight,
+  epochEndHeight,
+  gateways,
+}: {
+  epochStartHeight: BlockHeight;
+  epochEndHeight: BlockHeight;
+  gateways: DeepReadonly<Gateways>;
+}): Gateways {
+  const eligibleGateways: Gateways = {};
+  for (const [address, gateway] of Object.entries(gateways)) {
+    if (
+      isGatewayEligibleForDistribution({
+        epochStartHeight,
+        epochEndHeight,
+        gateway,
+      })
+    ) {
+      eligibleGateways[address] = gateway;
+    }
+  }
+  return eligibleGateways;
+}
+
+export async function getPrescribedObserversForEpoch({
+  gateways,
+  distributions,
+  minNetworkJoinStakeAmount,
+  epochStartHeight,
+  epochEndHeight,
+}: {
+  gateways: DeepReadonly<Gateways>;
+  distributions: DeepReadonly<RewardDistributions>;
+  minNetworkJoinStakeAmount: number;
+  epochStartHeight: BlockHeight;
+  epochEndHeight: BlockHeight;
+}): Promise<WeightedObserver[]> {
+  const weightedObservers: WeightedObserver[] = [];
+  let totalCompositeWeight = 0;
+
+  // filter out gateways eligible for epoch distribution
+  const eligibleGateways = getEligibleGatewaysForEpoch({
+    epochStartHeight,
+    epochEndHeight,
+    gateways,
+  });
+
+  // Get all eligible observers and assign weights
+  for (const [address, eligibleGateway] of Object.entries(eligibleGateways)) {
+    const stake = eligibleGateway.operatorStake;
+    const stakeWeight = stake / minNetworkJoinStakeAmount; // this has to be > 1 to joint he network
+
+    // the percentage of the epoch the gateway was joined for before this epoch
+    const totalBlocksForGateway =
+      epochStartHeight.valueOf() - eligibleGateway.start;
+    const calculatedTenureWeightForGateway =
+      totalBlocksForGateway / TENURE_WEIGHT_TOTAL_BLOCK_COUNT; // do not default to 0
+    const gatewayTenureWeight = Math.min(
+      calculatedTenureWeightForGateway,
+      MAX_TENURE_WEIGHT,
+    );
+
+    // the percentage of epochs participated in that the gateway passed
+    const totalEpochsParticipatedIn =
+      distributions.gateways[address]?.totalEpochParticipationCount || 0;
+    const totalEpochsGatewayPassed =
+      distributions.gateways[address]?.passedEpochCount || 0;
+    const gatewayRewardRatioWeight = totalEpochsParticipatedIn
+      ? totalEpochsGatewayPassed / totalEpochsParticipatedIn
+      : 1;
+
+    // the percentage of epochs the observer was prescribed and submitted reports for
+    const totalEpochsPrescribed =
+      distributions.observers[address]?.totalEpochsPrescribedCount || 0;
+    const totalEpochsSubmitted =
+      distributions.observers[address]?.submittedEpochCount || 0;
+    const observerRewardRatioWeight = totalEpochsPrescribed
+      ? totalEpochsSubmitted / totalEpochsPrescribed
+      : 1;
```

Comment:
these are using the values stored in the state for the gateway. they are updated each epoch.